Making gold files more readable and improves testing for the particle steppers

gsgall opened the pull request:
As @simopier requested in #23 I have transitioned the gold files for the single particle motion test cases away from utilizing exodus files for ray paths to csv files which show particle positions and velocities at each timestep. This increases the readability of the gold files and will help future developers/maintainer diagnose issues which may be introduced later on. 

- To generate the csv files the output block of the current test state was updated to produce the gold files.
- After the new gold csv files were generated based on the old input files all tests were updated to the current state.
- Verification of the stepper was re-confirmed and presented during the bi weekly technical meeting. 
- Additionally, common parts of the files were extracted to make use of the include input file syntax to reduce the amount of input file text that must be maintained. 

The test that was added was to improve the testing suite for the leap frog stepper and the part of the Boris stepper which accelerates particles through electric fields. This test case utilizes a time varying electric field that will produce the same force on the particle as the static magnetic field in the case of cyclotron motion. This enables demonstration of proper convergence order with the leapfrog stepper and the Boris stepper in the absence of a magnetic field.

gsgall committed:
Added specialization for a single particle data vector postprocessor

- This overrides the initialize function of the particle data vector
  postprocessor so that we can output all of the particle data over time
  to a single file

gsgall committed:
Refactoring tests to make the output more readable in line with #23

- Single particle motion tests now compare a single csv file output at
  the file step and fields

- fields are in tests to ensure consistency of the problem set up during
  testing

gsgall committed:
Refactored input files to reduce the amount of replicated inputs for the
test cases

- The boris stepper now has a base file for particle motion
	- The next step will be to refactor again and make it so that
	  there is a common file base between the boris and leapfrog
	  steppers

gsgall committed:
Improved gold files for leapfrog and refactored out a common input file

- Boris and leapfrog steppers now have a common file which their bases
  are built upon

- Leap frog stepper tests now also output csv files which are easy to
  read compared to ray exodus output files

gsgall committed:
Fixing source code comment blocks based on feedback from #97

simopier reviewed (changes requested):
Things look great overall, but I have two requests:

- [x] could you add documentation to the existing page about the new test?
- [x] could you add python scripts that can take the csv files and create figures to visualize the results of these tests, and add these figures to the documentation? See TMAP8 as an example for that.

gsgall committed:
Fixing the single particle motion VectorPostprocessor to work in parallel

gsgall committed:
Adding the requested information for the included input files in the
testing suite

- Added comments that describe the what the files which are included in
  other input files do

gsgall committed:
Fixing the single particle data processor and adding files to run
convergence tests

- The TestSingleParticleDataVectorPostprocessor did not work properly in
  parallel this fixes that bug

- Addintionally as PC requested the convergence studies have been
  added as heavy tests and figure generation will be added to the
  docuemtnation later

- A potnetial issue is still that the tests for covnergence are hard
  coded to run on one mpi processor but really they should be able to
  run on multiple and achieve the same results

- Proper documentation also needs to be added for the tests that were
  added and documentation that generates figures should also be added.

simopier commented on doc/config.yml:
```diff
@@ -88,25 +88,25 @@ Extensions:
+#                - /Cardinal/ICs/BulkEnergyConservation/BulkEnergyConservationICAction
+#                - /Cardinal/ICs/VolumetricHeatSource/VolumetricHeatSourceICAction
     MooseDocs.extensions.common:
         shortcuts: !include ${MOOSE_DIR}/framework/doc/globals.yml
-    MooseDocs.extensions.acronym:
-        acronyms:
-            framework: !include ${MOOSE_DIR}/framework/doc/acronyms.yml
-            cardinal: !include ${ROOT_DIR}/cardinal/doc/acronyms.yml
-            salamander:
-                SALAMANDER: Software for Advanced Large-scale Analysis of MAgnetic confinement for Numerical Design, Engineering & Research
-                PIC: particle-in-cell
-                CFD: computational fluid dynamics
-                CAD: computer-aided design
+    MooseDocs.extensions.acronym: disable
+#        acronyms:
+#            framework: !include ${MOOSE_DIR}/framework/doc/acronyms.yml
+#            cardinal: !include ${ROOT_DIR}/cardinal/doc/acronyms.yml
+#            salamander:
+#                SALAMANDER: Software for Advanced Large-scale Analysis of MAgnetic confinement for Numerical Design, Engineering & Research
+#                PIC: particle-in-cell
+#                CFD: computational fluid dynamics
+#                CAD: computer-aided design
```
Why do you need this? It seems like a loss to disable this.

gsgall replied:
This PR is still a WIP and this being disabled is only for local development of the docs so that I don't have to clone and build TMAP or Cardinal since I am not touching that part of the code. But the docs won't build properly without those applications built.

simopier commented on run-cases.py:
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env python
```
Add a short comment describing what this script does, and add short comments throughout to explain the script, or at least provide the main steps.

simopier commented on test/tests/userobjects/particle_stepper/analysis.py:
```diff
@@ -0,0 +1,134 @@
+import os
```
Please add a short comment explaining what this does.

simopier commented on test/tests/userobjects/particle_stepper/boris_stepper/test.py:
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env python
```
Please add a comment describing what this script does.

simopier commented on test/tests/userobjects/particle_stepper/boris_stepper/tests:
```diff
@@ -13,14 +13,38 @@
+    [convergence]
+      type = PythonUnitTest
+      input = 'test.py'
+      heavy = True
```
last time we talked, these tests did not even take that long. If they take 2 seconds or less in total, no need for heavy = true. If they take more than 4, let's leave heavy=true. If we're in the middle, let's chat, but probably leave heavy=true.

simopier commented on test/tests/userobjects/particle_stepper/boris_stepper/tests:
```diff
@@ -13,14 +13,38 @@
+    [convergence_errors]
+      type = 'CSVDiff'
+      input = 'test.py'
+      heavy = True
```
same here about heavy=true

simopier commented on test/tests/userobjects/particle_stepper/boris_stepper/tests:
```diff
@@ -13,14 +13,38 @@
+    [convergence_orders]
+      type = 'CSVDiff'
+      input = 'test.py'
+      heavy = True
```
same here about heavy=true

simopier commented on test/tests/userobjects/particle_stepper/boris_stepper/tests:
```diff
@@ -13,14 +13,38 @@
   []
   [cyclotron_motion]
     requirement = 'The system shall be capable of accurately capturing the path of charged particles in a perpendicular magnetic field'
+    [convergence]
```
I think you should have a `detail` section for each of these convergence tests.

simopier commented on test/tests/userobjects/particle_stepper/boris_stepper/tests:
```diff
@@ -31,14 +55,38 @@
   []
   [circular_e_field]
     requirement = 'The system shall be capable of accurately capturing the path of charged particles in a time varrying electric field'
+    [convergence]
+      type = PythonUnitTest
+      input = 'test.py'
+      heavy = True
+      test_case = BorisCyclotronElectricTest
+      required_python_packages = 'pandas matplotlib'
+      installation_type = in_tree
+    []
+    [convergence_errors]
+      type = 'CSVDiff'
+      input = 'test.py'
+      heavy = True
+      csvdiff = 'circular_e_field_errors.csv'
+      should_execute = False
+      prereq = 'circular_e_field/convergence'
+    []
+    [convergence_orders]
+      type = 'CSVDiff'
+      input = 'test.py'
+      heavy = True
+      csvdiff = 'circular_e_field_orders.csv'
+      should_execute = False
+      prereq = 'circular_e_field/convergence'
+    []
```
Same comments as with the other convergence tests above about `heavy = true` and the need for `detail`.

simopier commented on test/tests/userobjects/particle_stepper/leapfrog_stepper/circular_e_field.i:
```diff
@@ -1,3 +1,4 @@
+# This includes the stepper_base.i file and the tells the simulation to use the leapfrog stepper
```
```suggestion
# This includes the stepper_base.i file and tells the simulation to use the leapfrog stepper
```

simopier commented on test/tests/userobjects/particle_stepper/leapfrog_stepper/parallel_acceleration.i:
```diff
@@ -1,3 +1,4 @@
+# This includes the stepper_base.i file and the tells the simulation to use the leapfrog stepper
```
```suggestion
# This includes the stepper_base.i file and tells the simulation to use the leapfrog stepper
```

gsgall replied:
Used this suggestion.

simopier commented on test/tests/userobjects/particle_stepper/leapfrog_stepper/projectile_motion.i:
```diff
@@ -1,3 +1,4 @@
+# This includes the stepper_base.i file and the tells the simulation to use the leapfrog stepper
```
```suggestion
# This includes the stepper_base.i file and tells the simulation to use the leapfrog stepper
```

simopier commented on test/tests/userobjects/particle_stepper/leapfrog_stepper/test.py:
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env python
```
Add short comment.

simopier commented on test/tests/userobjects/particle_stepper/leapfrog_stepper/tests:
```diff
@@ -3,6 +3,30 @@
   issues = '#4 #19 #23'
   [circular_e_field]
     requirement = 'The system shall be capable of accurately capturing the path of charged particles in a time varrying electric field'
+    [convergence]
+      type = PythonUnitTest
+      input = 'test.py'
+      heavy = True
+      test_case = LeapFrogCyclotronElectricTest
+      required_python_packages = 'pandas matplotlib'
+      installation_type = in_tree
+    []
+    [convergence_errors]
+      type = 'CSVDiff'
+      input = 'test.py'
+      heavy = True
+      csvdiff = 'circular_e_field_errors.csv'
+      should_execute = False
+      prereq = 'circular_e_field/convergence'
+    []
+    [convergence_orders]
+      type = 'CSVDiff'
+      input = 'test.py'
+      heavy = True
+      csvdiff = 'circular_e_field_orders.csv'
+      should_execute = False
+      prereq = 'circular_e_field/convergence'
+    []
```
Same comment as with Boris for `detail` and `heavy = true`.

simopier reviewed (changes requested):
This is failing pre-check because of whitespaces. You need to set up your VSCode to avoid whitespaces.
In the meantime, for this PR, do: `
Run the "delete_trailing_whitespace.sh" script in your $MOOSE_DIR/scripts directory.`

I also left comments for you to consider throughout.

gsgall committed:
Saving current state

gsgall committed:
Adding more plotting infrastructure for verification

gsgall committed:
Adding the symlink for the plotting script

gsgall committed:
Cleaning up python scripts and add documentation for verification and in
scripts

- Added the verificiation documentation page for the stepper
  verification pages.
- Added function documentation and comments to all of the python scripts
  that are used during testing.
- Decided to keep the convergence studies as heavy tests since they took
  about 5-6 seconds when run on my machine.

gsgall committed:
Cleaning up some small issues

simopier commented on doc/content/verification_validation_examples/plasma/particle_stepping.md:
```diff
@@ -0,0 +1,142 @@
+For the sake of simplicity all particle properties are set to unity, in S.I. units, and the magnetic field magnitude $B_0$ is set to 1 T.
+Four time steps are used, each one an order of mangitude smaller than the last, $\Delta t \in [1, 10^{-1}, 10^{-2}, 10^{-3}]$ seconds.
+After a simulation at a given time step is completed the relative $l_2$, and $l_\infty$ norms of the error between the analytic solution and simulated are taken.
+Plotting these errors on a log-log plot and then taking the slope reveals the order of accuracy of the time integration, both implemented steppers are second order accurate in time, so the slope of each line should be two.
```
```suggestion
Plotting these errors on a log-log plot and then deriving the slope reveals the order of accuracy of the time integration, both implemented steppers are second order accurate in time, so the slope of each line should be two.
```

simopier commented on doc/content/verification_validation_examples/plasma/particle_stepping.md:
```diff
@@ -0,0 +1,142 @@
+       caption=Relative $l_\infty$ error for particle position and velocity for a particle being moved in a time varrying electric field via the Leapfrog Stepper for four different time step sizes.
+
+!col-end!
+!row-end!
```
Add a few sentences discussing the results, and stating if the verification is sucessful or not. Make sure to cross reference al of the figures (as always, every figure needs to e referenced and discussed in the text).

simopier replied:
This was not resolved since the cross-references were not active ones.

simopier commented on doc/content/verification_validation_examples/index.md:
```diff
@@ -19,9 +19,10 @@ SALAMANDER couples with other MOOSE-based applications, such as [TMAP8](https://
-However, several verification cases are already available in the `salamander/test/tests` folder.
+| Case    | Title                                                                              |
+| ------- | ---------------------------------------------------------------------------------- |
+| 1 | [Particle Stepper](particle_stepping.md) |
```
```suggestion
| 1 | [Particle in Cell - Particle Stepper](particle_stepping.md) |
```

simopier reviewed (changes requested):
It looks like you mistakenly pushed the moose submodule, which is why your tests are not running. 

This is more of a note to myself: Since the documentation does not build right now, I can't review its rendering. I'll need to review the documentation pages once it builds (including figures).

gsgall committed:
Cleaning up documentation

- More detail has been added to the documentation of the verificaiton
  case for the particle stepper.
- Documentation also now automatically generates the figures needed for
  these verification cases.

gsgall committed:
Removing trailing white spaces for CI

gsgall committed:
Updating test details and requirements in order to meet SQA requirements

gsgall committed:
Going back and addressing missed comments

- I missed a few comments in the PR that requested changes or additional
  documentation so I have gone back and taken care of these

- Mostly minimal changes that add additional documentation where
  requested

gsgall commented on moose:
I have removed the submodule update

simopier commented on test/tests/userobjects/particle_stepper/stepper_analysis.py:
```diff
@@ -0,0 +1,228 @@
+steppers.
+The test case is designed around the canonical example of single particle motion, cyclotron motion.
+The functions in this file provide everything required to calculate the analytic solution and compare
+the actual path provided by the system to the analytic.
```
```suggestion
the actual path provided by the system to the analytical solution.
```

simopier commented on doc/content/verification_validation_examples/plasma/particle_stepping.md:
```diff
@@ -0,0 +1,159 @@
+
+# Demonstration Problem
+
+The example of single particle motion utilized in the verification of these capabilities is with a classic examples of single particle motion, cyclotron motion.
```
```suggestion
The verification of the single particle motion use classic examples of single particle motion and cyclotron motion.
```

simopier commented on doc/content/verification_validation_examples/plasma/particle_stepping.md:
```diff
@@ -0,0 +1,159 @@
+# Demonstration Problem
+
+The example of single particle motion utilized in the verification of these capabilities is with a classic examples of single particle motion, cyclotron motion.
+While cycltron motion requires a magnetic field we will utilize the analytic solution for particle motion to configure electric fields to reproduce the same praticle paths.
```
```suggestion
While cyclotron motion requires a magnetic field, we will utilize the analytic solution for particle motion to configure electric fields to reproduce the same particle paths.
```

simopier commented on doc/content/verification_validation_examples/plasma/particle_stepping.md:
```diff
@@ -0,0 +1,159 @@
+Within SALAMANDER two main particle steppers are provided, the [Leapfrog](LeapFrogStepper.md) stepper and the [Boris](BorisStepper.md) stepper. Both of these particle steppers are second order accurate in time.
+In order to verify that these steppers work properly a convergence studies can be utilized to ensure that second order accuracy is observed.
+
+# Demonstration Problem
```
```suggestion
# Demonstration Problems
```

simopier commented on doc/content/verification_validation_examples/plasma/particle_stepping.md:
```diff
@@ -0,0 +1,159 @@
+  \vec{v} \times \vec{B}
+\end{equation}
+
+Where $\vec{v}$ is the particle's velocity, $q$ is the particle's charge, and $m$ is the particle's mass.
```
```suggestion
where $\vec{v}$ is the particle's velocity, $q$ is the particle's charge, and $m$ is the particle's mass.
```

simopier commented on doc/content/verification_validation_examples/plasma/particle_stepping.md:
```diff
@@ -0,0 +1,159 @@
+Where $r_L$ is the what is called the Larmor radius and given by
+
+\begin{equation}
+  r_L = \frac{v_\perp}{\omega_c}
```
```suggestion
  r_L = \frac{v_\perp}{\omega_c},
```

simopier commented on doc/content/verification_validation_examples/plasma/particle_stepping.md:
```diff
@@ -0,0 +1,159 @@
+\end{equation}
+
+We can then plug in the solution for the particles path from the example of cycltron motion to calculate the electric field which will reproduce the same particle motion.
+Doing so yields the electric field
```
```suggestion
Doing so yields the electric field:
```

simopier commented on doc/content/verification_validation_examples/plasma/particle_stepping.md:
```diff
@@ -0,0 +1,159 @@
+## Convergence Studies
+
+For the sake of simplicity, all particle properties are set to unity, in S.I. units, and the magnetic field magnitude $B_0$ is set to 1 T.
+Four time steps are used, each one an order of mangitude smaller than the last, $\Delta t \in [1, 10^{-1}, 10^{-2}, 10^{-3}]$ seconds.
```
```suggestion
Four time steps are used, each one an order of magnitude smaller than the last, $\Delta t \in [1, 10^{-1}, 10^{-2}, 10^{-3}]$ seconds.
```

simopier commented on doc/content/verification_validation_examples/plasma/particle_stepping.md:
```diff
@@ -0,0 +1,159 @@
+### Boris Stepper: Cyclotron Motion
+
+This test specifically tests the part of the stepper which calculates the force on the particle due to the magnetic field.
+In examing the errors from the figures below it can be seen that the implementatio of the Boris Stepper utilized achieves the expected second order accuracy in time for both $x$, and $y$ particle positions as well as the $x$, and $y$ componenets of the particles velocity.
```
```suggestion
In examining the errors from the figures below it can be seen that the implementation of the Boris Stepper achieves the expected second order accuracy in time for both $x$ and $y$ particle positions as well as the $x$ and $y$ components of the particles velocity.
```

simopier commented on doc/content/verification_validation_examples/plasma/particle_stepping.md:
```diff
@@ -0,0 +1,159 @@
+
+Since the Leapfrog stepper can be thought of as a special case of the Boris stepper where no magentic field is present the resulting errors should be the same as those produced in the previous test.
+If the figures and calculated orders of accuracy in the above section are compared with the figures and orders of accuracy in this section they should be the same.
+In doing this comparison it is clear that this is the case and the two steppers, as expected, produce the same results.
```
```suggestion
In doing this comparison, it is clear that this is the case and the two steppers, as expected, produce the same results.
```

simopier replied:
Cross reference the relevant figures.

simopier reviewed (changes requested):
Thank you @gsgall, this is getting very close. I still have some minor comments and suggestions for you, let me know if you have any questions.

gsgall committed:
Apply suggestions from code review

Co-authored-by: Pierre-Clement Simon <pierreclement.simon@gmail.com>

simopier commented on doc/content/verification_validation_examples/plasma/particle_stepping.md:
Still to be done.

gsgall committed:
Addressing final comments

simopier commented on test/tests/userobjects/particle_stepper/stepper_analysis.py:
```diff
@@ -0,0 +1,228 @@
+
+
+def analytic_v_x(v_perp: float, m: float, q: float, B: float, t: np.ndarray):
+    """ Computes the analytic solution for x component of a particles velocity subject to cyclotron motion given a positive magnetic along the z diretion
```
```suggestion
    """ Computes the analytic solution for x component of a particles velocity subject to cyclotron motion given a positive magnetic along the z direction
```

simopier commented on test/tests/userobjects/particle_stepper/stepper_analysis.py:
```diff
@@ -0,0 +1,228 @@
+
+
+def analytic_v_y(v_perp: float, m: float, q: float, B: float, t: np.ndarray):
+    """ Computes the analytic solution for y component of a particles velocity subject to cyclotron motion given a positive magnetic along the z diretion
```
```suggestion
    """ Computes the analytic solution for y component of a particles velocity subject to cyclotron motion given a positive magnetic along the z direction
```

simopier reviewed (changes requested):
Some very minor things, but this is ready for @cticenhour to review

gsgall committed:
Apply suggestions from code review

Co-authored-by: Pierre-Clement Simon <pierreclement.simon@gmail.com>

simopier reviewed (dismissed):
Looks good to me. Let's make sure @cticenhour gets to review it too.

cticenhour commented on doc/content/verification_validation_examples/plasma/particle_stepping.md:
```diff
@@ -0,0 +1,185 @@
+- [/leapfrog_base.i] selects the [LeapFrogStepper.md] as the particle stepper.
+- [/boris_stepper/cyclotron_motion.i], [/boris_stepper/circular_e_field.i], and [/leapfrog_stepper/circular_e_field.i] setup the initial conditions for the particles and fields for each case.
+
+To combine them into one input file when running the simulation, the `!include` feature is utilized.
```
```suggestion
To combine them into one input file when running the simulation, the `!include` feature within MOOSE is utilized.
```

cticenhour commented on test/tests/userobjects/particle_stepper/boris_stepper/test.py:
```diff
@@ -0,0 +1,42 @@
+    This test case tests the Boris stepper using a time varrying electric field that
+    is spatially uniform.
+    The exact form of the electric field is calculated by propagating the solution for
+    the particles path that is provided by the cyclotron motion case through Newtons
```
```suggestion
    the particle's path that is provided by the cyclotron motion case through Newton's
```

cticenhour commented on test/tests/userobjects/particle_stepper/boris_stepper/tests:
```diff
@@ -1,75 +1,152 @@
+      test_case = BorisCyclotronMagneticTest
+      required_python_packages = 'pandas matplotlib'
+      installation_type = in_tree
+      detail = 'and perform a convergence study for particle motion'
```
```suggestion
      detail = 'performing a convergence study for particle motion,'
```

cticenhour commented on test/tests/userobjects/particle_stepper/boris_stepper/tests:
```diff
@@ -1,75 +1,152 @@
+      csvdiff = 'circular_e_field_errors.csv'
+      should_execute = False
+      prereq = 'circular_e_field/convergence'
+      detail = 'and ensure that errors are consistent'
```
```suggestion
      detail = 'ensuring that errors are consistent,'
```

cticenhour commented on test/tests/userobjects/particle_stepper/boris_stepper/tests:
```diff
@@ -1,75 +1,152 @@
+      csvdiff = 'circular_e_field_orders.csv'
+      should_execute = False
+      prereq = 'circular_e_field/convergence'
+      detail = 'and ensure that the stepper demonstrates the proper order of accuracy'
```
```suggestion
      detail = 'ensuring that the stepper demonstrates the proper order of accuracy,'
```

cticenhour commented on test/tests/userobjects/particle_stepper/boris_stepper/tests:
```diff
@@ -1,75 +1,152 @@
-               projectile_motion_rays.e-s0008
-               projectile_motion_rays.e-s0009'
-    allow_test_objects = true
     requirement = 'The system shall be capable of applying a linear impluse from a force field perpendicular to a particles initial velocity using the boris stepper when there is 0 magnetic field'
```
```suggestion
    requirement = 'The system shall be capable of applying a linear impluse from a force field perpendicular to a particle's initial velocity using the Boris stepper when there is zero magnetic field,'
```

cticenhour commented on test/tests/userobjects/particle_stepper/stepper_analysis.py:
```diff
@@ -0,0 +1,228 @@
+
+
+def analytic_x(v_perp: float, m: float, q: float, B: float, t: np.ndarray) -> np.ndarray:
+    """ Computes the analytic solution for x component of a particles position subject to cyclotron motion given a positive magnetic along the z direction
```
```suggestion
    """ Computes the analytic solution for x component of a particle's position subject to cyclotron motion given a positive magnetic field along the z direction
```

cticenhour reviewed (changes requested):
Note that I would like to do another review of the documentation portion when these are accepted. This looks good in general - most if not all of my comments are documentation tilted.

cticenhour reviewed (changes requested):
Note that I would like to do another review of the documentation portion when these are accepted. This looks good in general - most if not all of my comments are documentation tilted.

gsgall committed:
Apply suggestions from code review

Co-authored-by: Casey Icenhour <cticenhour@gmail.com>

simopier commented:
@gsgall some tests are failing with this error: 

```
/tmp/build/salamander/test/tests/userobjects/particle_stepper/leapfrog_stepper/tests:6: Failed to create Tester: The following parameters are deprecated and should use the 'capability' param instead; 'installation_type' -> capability 'installation_type'
```

You can check out https://github.com/idaholab/moose/pull/32294 and https://github.com/idaholab/TMAP8/pull/362 for guidance and examples on how to address this.

loganharbour commented on test/include/vectorpostprocessors/TestSingleParticleDataVectorPostprocessor.h:
```diff
@@ -0,0 +1,29 @@
+//* This file is part of SALAMANDER: Software for Advanced Large-scale Analysis of MAgnetic
```
double header

loganharbour commented on test/src/vectorpostprocessors/TestSingleParticleDataVectorPostprocessor.C:
```diff
@@ -0,0 +1,56 @@
+//* This file is part of SALAMANDER: Software for Advanced Large-scale Analysis of MAgnetic
```
double header

gsgall committed:
Updating the test specs to use the proper capabailities for tests

gsgall committed:
Addressing the majority of the formatting requests

- Need to go back and rewrite some of the comments for Casey's comments.
- Need to go back and check the input file links in Logan's comments.

gsgall committed:
Adding more detailed comments to the input file includes

- The comments now describe what the included file does.
- The comment also now describes what the rest of the input file does as
  well.

cticenhour commented on test/include/vectorpostprocessors/TestSingleParticleDataVectorPostprocessor.h:
```diff
@@ -0,0 +1,29 @@
+//* This file is part of SALAMANDER: Software for Advanced Large-scale Analysis of MAgnetic
+// confinement for Numerical Design, Engineering & Research,
```
```suggestion
//* confinement for Numerical Design, Engineering & Research,
```

cticenhour reviewed (changes requested):
Only one minor edit that should be addresses for consistency.

gsgall committed:
Fixing up header comment

- My auto formatter made some of the header comment wrap onto a new line
  it used // to start instead of //* and this just fixes that

moosebuild commented:
Job [Build test summary](https://civet.inl.gov/job/3710358/), step Build test summary on 7d05de7 wanted to post the following:

## Test summary

Compared against 64f1fee in job [civet.inl.gov/job/3709540](https://civet.inl.gov/job/3709540).

### Removed tests

| Test                                                                       | Time (s)   | Memory (MB)   |
|----------------------------------------------------------------------------|------------|---------------|
| `test:userobjects/particle_stepper/boris_stepper.parallel_acceleration`    | 0.48       | 179.06        |
| `test:userobjects/particle_stepper/leapfrog_stepper.parallel_acceleration` | 0.44       | 37.42         |
| `test:userobjects/particle_stepper/boris_stepper.e_cross_b`                | 0.41       | 64.55         |
| `test:userobjects/particle_stepper/boris_stepper.projectile_motion`        | 0.37       | 45.43         |
| `test:userobjects/particle_stepper/boris_stepper.cyclotron_motion`         | 0.37       | 50.27         |
| `test:userobjects/particle_stepper/leapfrog_stepper.projectile_motion`     | 0.37       | 12.91         |

### Added tests

| Test                                                                                     | Time (s)   | Memory (MB)   |
|------------------------------------------------------------------------------------------|------------|---------------|
| `test:userobjects/particle_stepper/boris_stepper.parallel_acceleration/fields`           | 0.57       | 88.33         |
| `test:userobjects/particle_stepper/leapfrog_stepper.parallel_acceleration/fields`        | 0.49       | 68.15         |
| `test:userobjects/particle_stepper/boris_stepper.e_cross_b/fields`                       | 0.45       | 22.87         |
| `test:userobjects/particle_stepper/leapfrog_stepper.circular_e_field/fields`             | 0.43       | 46.23         |
| `test:userobjects/particle_stepper/boris_stepper.circular_e_field/fields`                | 0.42       | 87.68         |
| `test:userobjects/particle_stepper/boris_stepper.cyclotron_motion/fields`                | 0.42       | 42.08         |
| `test:userobjects/particle_stepper/boris_stepper.projectile_motion/fields`               | 0.42       | 65.85         |
| `test:userobjects/particle_stepper/leapfrog_stepper.projectile_motion/fields`            | 0.40       | 41.17         |
| `test:userobjects/particle_stepper/boris_stepper.cyclotron_motion/convergence_orders`    | SKIP       | 0.00          |
| `test:userobjects/particle_stepper/boris_stepper.circular_e_field/convergence_orders`    | SKIP       | 0.00          |
| `test:userobjects/particle_stepper/boris_stepper.cyclotron_motion/convergence`           | SKIP       | 0.00          |
| `test:userobjects/particle_stepper/boris_stepper.circular_e_field/convergence`           | SKIP       | 0.00          |
| `test:userobjects/particle_stepper/leapfrog_stepper.circular_e_field/convergence_orders` | SKIP       | 0.00          |
| `test:userobjects/particle_stepper/boris_stepper.circular_e_field/convergence_errors`    | SKIP       | 0.00          |
| `test:userobjects/particle_stepper/leapfrog_stepper.circular_e_field/convergence`        | SKIP       | 0.00          |
| `test:userobjects/particle_stepper/leapfrog_stepper.circular_e_field/convergence_errors` | SKIP       | 0.00          |
| `test:userobjects/particle_stepper/boris_stepper.cyclotron_motion/convergence_errors`    | SKIP       | 0.00          |
| `test:userobjects/particle_stepper/boris_stepper.cyclotron_motion/particle`              |            | 0.00          |
| `test:userobjects/particle_stepper/leapfrog_stepper.parallel_acceleration/particle`      |            | 0.00          |
| `test:userobjects/particle_stepper/boris_stepper.circular_e_field/particle`              |            | 0.00          |
| `test:userobjects/particle_stepper/leapfrog_stepper.projectile_motion/particle`          |            | 0.00          |
| `test:userobjects/particle_stepper/boris_stepper.e_cross_b/particle`                     |            | 0.00          |
| `test:userobjects/particle_stepper/leapfrog_stepper.circular_e_field/particle`           |            | 0.00          |
| `test:userobjects/particle_stepper/boris_stepper.parallel_acceleration/particle`         |            | 0.00          |
| `test:userobjects/particle_stepper/boris_stepper.projectile_motion/particle`             |            | 0.00          |

cticenhour reviewed (dismissed):
Comments addressed. Looks good!

loganharbour commented on test/tests/userobjects/particle_stepper/leapfrog_stepper/tests:
```diff
@@ -1,37 +1,86 @@
+      input = 'test.py'
+      heavy = True
+      test_case = LeapFrogTimeVarryingElectricTest
+      required_python_packages = 'pandas matplotlib'
```
```suggestion
```

This is an implicit requirement on `MMSTest` already

loganharbour commented on test/tests/userobjects/particle_stepper/leapfrog_stepper/tests:
```diff
@@ -1,37 +1,86 @@
+      heavy = True
+      test_case = LeapFrogTimeVarryingElectricTest
+      required_python_packages = 'pandas matplotlib'
+      capabilities = 'installation_type=in_tree'
```
```suggestion
```

And I believe this is an implicit requirement as well. I removed it from all of the MOOSE tests and I can't figure out why but I believe you're good. In the future if we can actually enable MMS tests out of tree, it would be nice to see this fail here so we can see what we need to do to fix it

loganharbour commented on test/tests/userobjects/particle_stepper/boris_stepper/tests:
```diff
@@ -1,75 +1,152 @@
+      heavy = True
+      test_case = BorisCyclotronMagneticTest
+      required_python_packages = 'pandas matplotlib'
+      capabilities = 'installation_type=in_tree'
```
```suggestion
```

loganharbour commented on test/tests/userobjects/particle_stepper/boris_stepper/tests:
```diff
@@ -1,75 +1,152 @@
+      heavy = True
+      capabilities = 'installation_type=in_tree'
+      test_case = BorisTimeVarryingElectricTest
+      required_python_packages = 'pandas matplotlib'
```
```suggestion
      test_case = BorisTimeVarryingElectricTest
```

gsgall committed:
Apply suggestions from code review

Co-authored-by: Logan Harbour <loganhharbour@gmail.com>

simopier reviewed (approved):
Looks good! Thank you @gsgall for the contribution and @cticenhour and @loganharbour for the reviews!
